Wikipedia/Code/WMFAppViewController.m

```diff
@@ -189,14 +208,56 @@ - (void)pauseApp {
     [self performHousekeepingIfNecessary];
 }
 
+- (void)handleSelectionOfShortcutItemIfNecessary {
+    UIApplicationShortcutItem* shortcutItemSelectedAtLaunch = self.shortcutItemSelectedAtLaunch;
+    self.shortcutItemSelectedAtLaunch = nil;
+    if (shortcutItemSelectedAtLaunch) {
+        @weakify(self)
+        void (^ handleSelection)() = ^void () {
+            @strongify(self)
+            if ([shortcutItemSelectedAtLaunch.type isEqualToString:WMFIconShortcutTypeSearch]) {
+                // Show search without loosing underlying view hierarchy.
+
+                // HAX: fragile! fix!
+                UIBarButtonItem* searchButton = ((WMFArticleContainerViewController*)((UINavigationController*)self.rootTabBarController.selectedViewController).topViewController).navigationItem.rightBarButtonItem;
+                NSAssert(searchButton, @"Search button not found.");
+                NSAssert([searchButton.target respondsToSelector:searchButton.action], @"Search button expected selector not found.");
+
+                [searchButton.target performSelector:searchButton.action withObject:searchButton afterDelay:0];
+            } else if ([shortcutItemSelectedAtLaunch.type isEqualToString:WMFIconShortcutTypeRandom]) {
+                [self.tabBarController setSelectedIndex:WMFAppTabTypeHome];
+                [[self navigationControllerForTab:WMFAppTabTypeHome] popToRootViewControllerAnimated:NO];
+                [self showRandomArticleAnimated:YES];
+            } else if ([shortcutItemSelectedAtLaunch.type isEqualToString:WMFIconShortcutTypeNearby]) {
+                [self.tabBarController setSelectedIndex:WMFAppTabTypeHome];
+                [[self navigationControllerForTab:WMFAppTabTypeHome] popToRootViewControllerAnimated:NO];
+                [self showNearbyListAnimated:YES];
+            } else if ([shortcutItemSelectedAtLaunch.type isEqualToString:WMFIconShortcutTypeContinueReading]) {
+                // Fall through to normal "Continue reading..." handling.
```

Comment:
@montehurd also note that if it has been more than 24 hours, the article won't actually be pushed (it will just be the first card on the home screen)